commons/src/gem_version.rs

```diff
@@ -0,0 +1,230 @@
+use fancy_regex::Regex;
+use std::cmp;
+use std::cmp::Ordering;
+use std::fmt;
+use std::str::FromStr; // needed for lookahead/behind
+
+/// # Struct to hold semver-ish versions for comparison
+///
+/// Based off of Ruby's `Gem::Version` logic:
+///
+/// - <https://github.com/rubygems/rubygems/blob/ecc8e895b69063562b9bf749b353948e051e4171/lib/rubygems/version.rb>
+/// - <https://github.com/rubygems/rubygems/blob/ecc8e895b69063562b9bf749b353948e051e4171/test/rubygems/test_gem_version.rb>
+///
+/// Example:
+///
+/// ```rust
+/// use std::str::FromStr;
+/// use commons::gem_version::GemVersion;
+///
+/// let version = GemVersion::from_str("1.0.0").unwrap();
+/// assert!(version < GemVersion::from_str("2.0.0").unwrap());
+/// ```
+#[derive(Debug, Default)]
+pub struct GemVersion {
+    segments: Vec<VersionSegment>,
+}
+
+impl fmt::Display for GemVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let version_string = self
+            .segments
+            .iter()
+            .map(|segment| match segment {
+                VersionSegment::String(s) => s.clone(),
+                VersionSegment::U32(i) => i.to_string(),
+            })
+            .collect::<Vec<String>>()
+            .join(".");
+        write!(f, "{}", version_string)
+    }
+}
+
+impl FromStr for GemVersion {
+    type Err = VersionError;
+
+    fn from_str(version_string: &str) -> Result<Self, Self::Err> {
+        if version_string.trim().is_empty() {
+            Ok(GemVersion {
+                segments: vec![VersionSegment::U32(0)],
+            })
+        } else {
+            let validation_regex = Regex::new(
+                "\\A\\s*([0-9]+(?>\\.[0-9a-zA-Z]+)*(-[0-9A-Za-z-]+(\\.[0-9A-Za-z-]+)*)?)?\\s*\\z",
+            )
+            .expect("Internal Error: Invalid Regular Expression!");
+
+            let segment_regex =
+                Regex::new("[0-9]+|[a-z]+").expect("Internal Error: Invalid Regular Expression!");
+
+            if validation_regex.is_match(version_string).unwrap_or(false) {
+                let (segments_l, segments_r) = segment_regex
+                    .find_iter(version_string)
+                    .map(|regex_match| {
+                        let match_string = String::from(regex_match.unwrap().as_str());
```

Comment:
This is an unwrap in non-test code? Can this ever happen? Should it be an `.expect()` instead to be self-documenting and/or could this be refactored to avoid the unwrap if it's provable not going to occur?

Reply:
This review comment isn't resolved

Reply:
Updated to remove the unwrap. This is only needed for `fancy_regex` however we have both that and `regex` loaded, so I switched to using it here